### ISSUE TITLE
MySQLAdvanced Datastore ConnectionPool hang

### ISSUE THREAD

**Issue by pidydx** [google/grr#170]
We are seeing workers that are still running, but appear to stop doing work.  The last error before they appear to drop off of the Server Load active thread count is:

Trace: Traceback (most recent call last):
  File "/opt/grr/lib64/python2.7/site-packages/grr-0.3.0_6-py2.7.egg/grr/lib/flow_runner.py", line 548, in RunStateMethod
    responses=responses)
  File "/opt/grr/lib64/python2.7/site-packages/grr-0.3.0_6-py2.7.egg/grr/lib/flow.py", line 312, in Decorated
    auth_required=auth_required)
  File "/opt/grr/lib64/python2.7/site-packages/grr-0.3.0_6-py2.7.egg/grr/lib/flow.py", line 149, in __init__
    raise FlowError("No valid Status message.")
FlowError: No valid Status message.

This appears to be associated primarily Interrogate hunts and 3.0.0.3 clients.

**Comment by pidydx** [comment]
I put a worker in debug mode and let it run for about an hour or two while busy and it eventually just stopped with this message.
DEBUG:2015-05-08 17:41:09,951 stats_store:736] Writing stats to stats store.

It had written to the stats store multiple times before it got hung up here and it doesn't give any errors and the process doesn't die.  Looking through the only thing I see that might be related to an infinite hang is something going wrong in the loops in stats_store.py around line 169 in WriteStats()

**Comment by pidydx** [comment]
I added a bunch of debug statements and it seems to hang here in stats_store.py:L243 in WriteStats.
    process_data = aff4.FACTORY.Create(self.urn.Add(process_id),
                                       "StatsStoreProcessData",
                                       mode="rw", token=self.token)

I put debugging statements in aff4.FACTORY.Create and it hasn't reproduced in 12+ hrs now.  I don't see anything special here so I suspect that GRR is occasionally hanging up on this call elsewhere, but the stats store one is what hangs the workers.

Until I can get more information deeper in the error I suspect this may be a rare timing issue in MySQLAdvanced datastore, but I'm not clear what could cause aff4.FACTORY.Create to permanently hang rather than raise and error somewhere.  Any thoughts on what could cause this or where to look would be greatly appreciated.  Thanks!

**Comment by pidydx** [comment]
Good news.  I found the bug (mostly).  It is in the MySQLAdvanced connector, but thankfully it does not appear to be related to any of the MySQL stuff directly.

Bad News.  It is the ConnectionPool object.  In trying to debug this I found a significant performance problem related to the pool not being able to handle so many threads so I modified it a bit to try to grow/shrink as necessary. https://github.com/pidydx/grr/tree/SQLPoolFix.  However, in both the original and the updated version after a period of time the ConnectionPool object's self.connections = Queue.Queue() attribute just dies.  Even just trying to do self.connections.qsize() results in an infinite hang when this condition occurs.  Unfortunately, I am not clear as to what is causing this condition or how to avoid it though.
